### PR TITLE
Fix memory leak in FA assembly

### PR DESCRIPTION
**Pull request by pazner**
The `FULL` assembly mode has a memory leak since the initial pointers for the `I`, `J`, and `A` arrays are overwritten (`J` and `A` are allocated but are actually zero-length, `I` has nonzero length).

This PR adds a new "expert function" `SparseMatrix::OverrideSize` that lets one change the `width` and `height` of a `SparseMatrix` after it has already been constructed (e.g. with the default constructor).

The use-case here is creating an "empty" `SparseMatrix` (but with the correct dimensions), and letting the user handle allocating and filling the CSR arrays.

### CODE REVIEW

**Review by YohannDudouit — APPROVED** [review]
I remember struggling finding a constructor that wouldn't do or allocate things I don't want, ideally the `SparseMatrix::SparseMatrix(int nrows, int ncols)` constructor should just do that I think. I think this issue is arising because this `SparseMatrix` class conflates two sparse matrix representations into one. I think the code would gain in clarity and simplicity by separating the `SparseMatrix` class in two, one based on linked list (`LLSparseMatrix` for instance), and one based on the CSR format (`CSRSparseMatrix` for instance). By conflating these two sparse matrix representations into one, it is hard to follow what is actually happening inside the current `SparseMatrix` class implementation, and the interface is hard to use correctly because impacting both these representations internally at the same time.

**Review by rcarson3 — APPROVED** [review]
I guess this seems fine with me after looking over how the class does things.

**Comment by tzanio** [comment]
Merged in `next` for testing...
